### ISSUE TITLE
incorrect formatting of JSDoc type assertion

### ISSUE THREAD

**Issue by bradzacher** [prettier/prettier#8171]
**Prettier 2.0.5**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEATOYA2BDATnAAkigGcYCAzCCJA7KATwG4AdKN4sggIzwEYCAXgIB6AFRiCAARgMADoWBlcASygBzAL4ExIgBRUIASiYgANCAhyYK6CWSg8uCAHcACngT2U2TC+wM9hbcuNhgANZwMADKcmFq6sgwuACucBYAFjAAtpgA6hkq8CRxYHDRXkUqAG5FDMjgJEEgaiRwuDBuoerZ2MgUvm0WAFYkAB4AQqERUdHY2XAAMmpw-YPpIKNj0QmYcACKKRDwa5hDIHG4bbgNsgokYKrW5heqsHkqqDAZyAAcAAwWOTONp5UJyBrAuDXaqrCwARyO8C6Vm8IGwJAAtFA4HB0KgXvhESp8F1sD0+kgBmcNm1siokqlabsDkjVlT1hYYNhuB8vj8kAAmLmhFSYBIAYQg2V6DWhAFYXik2gAVHneannappACSUHQsGijxU1gAgvrorI9qc2ppNEA)

```sh
# Options (if any):
--parser typescript
```

**Input:**

```tsx
declare const foo: any;

const bar1 = /** @type {string} */(foo);
```

**Output:**

```tsx
declare const foo: any;

const bar1 = /** @type {string} */ foo;
```

**Expected behavior:**

```tsx
declare const foo: any;

const bar1 = /** @type {string} */(foo);
// or
const bar1 = /** @type {string} */ (foo);
```

### What's wrong?

In a JS Codebase, dropping the parens breaks the JSDoc type assertion, which will break type checking.

```ts
declare const foo: any;

const bar1 = /** @type {string} */(foo);
// typeof bar1 === string

const bar2 = /** @type {string} */ foo;
// typeof bar2 === any
```

See deeper explanation in https://github.com/typescript-eslint/typescript-eslint/issues/1943

It looks like this only occurs with the typescript parser, as [the babel parser](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEAjAhgJwEZMBeTAegCpLMABGATwAc5NgN8BLKAcwF9MlcgAoAZhAgBKANwgANCAhMYndMlAF8EAO4AFAgjTIQuADbbcDIwuz5cYANZwYAZSb3uPZDHwBXOAoAFjAAtqYA6oGc8GjuYHAuhtGcAG7RDMZgaNYg3Ghw+DC6djwhuMiiZvkKAFZoAB4AQnaOzi64IXAAMtxwFVUBIHX1Lp6mcACKvhDw-abVIO74+fjGeNhwpvKLXLDhnAAmMIHIABwADApMWvnhdkzG13ArKX0KAI7T8MVKRii4aAAtFA4HADmDtvg4J9OFDirhSuUkJV5oN8iFON4-GixpMvn1kQMFDBcNh9kcTkgAEzEuycUyeADCEBCZWMzwArNtfPkACqkv4ohYpfwASSg4NgLjAXGUAEEJS5GOM5vk+HwgA) works correctly (though note that the `flow` parser exhibits the same paren scrubbing behaviour).

Related issues (not exactly the same, but pertaining to JSDoc in TS): #8045, #5850

**Comment by thorn0** [comment]
It's a known issue. See the [release notes](https://prettier.io/blog/2020/03/21/2.0.0.html#ultimate-fix-for-closure-style-type-casts-7791httpsgithubcomprettierprettierpull7791-by-thorn0httpsgithubcomthorn0-7011httpsgithubcomprettierprettierpull7011-by-evilebottnawihttpsgithubcomevilebottnawi) for 2.0. TypeScript the language doesn't support such type casts. TypeScript the compiler supports them only in JS.

**Comment by thorn0** [comment]
People might have reasons to use `--parser typescript` to format JS, but still I wouldn't call this a bug. If the parser supported `ParenthesizedExpression`, it'd work for TS, but it doesn't seem to be a big deal that it doesn't.

**Comment by bradzacher** [comment]
No problem! I was just raising it because of the attached issue that someone raised - in testing I realised that prettier munges it.

I think we might need to add support for at least this case to the parser, which would fix the issue here as well.

**Comment by thorn0** [comment]
I didn't know the `@typescript-eslint/*` rules are used for JS.

BTW, it'd be interesting if the parser exposed the `jsDoc` property or even converted the types in it to ESTree-like nodes.

**Comment by bradzacher** [comment]
Yeah (almost) all of our rules work fine on JS code. If you're good with your jsdoc typing, ones that use type information will even work pretty well!

Unfortunately it looks like JSDoc will never be an official part of the ESTree spec (https://github.com/estree/estree/issues/201), but there's no reason we couldn't _consider_ exposing it, given that it's parsed as an AST by TS itself.
Would be an interesting experiment

**Comment by hosseinmd** [comment]
I have a plugin for prettier formatting jsdoc. https://github.com/hosseinmd/prettier-plugin-jsdoc

**Comment by thorn0** [comment]
Related issue in the Meriyah repo: https://github.com/meriyah/meriyah/issues/183
